src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,241 @@
+
+// check if std::optional<V> or V is null using the same syntax - needed for templating
+template <typename T>
+constexpr bool is_null(const T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return !maybe_optional.has_value();
+  } else {
+    return false;
+  }
+}
+
+// get the value of std::optional<V> or V using the same syntax - needed for templating
+template <typename T>
+get_value_type<T>& get_value(T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return maybe_optional.value();
+  } else {
+    return maybe_optional;
+  }
+}
+
+}  // namespace table_builder
+
+/**
+ * Helper to build a table with a static column layout, specified by constructor args types and names. Keeps a
+ * value vector for each column and appends values to them in append_row(). For nullable columns an additional
+ * null_values vector is kept. Automatically creates chunks in accordance with the specified chunk size.
+ */
+template <typename... DataTypes>
+class TableBuilder {
+ public:
+  // names is a boost::hana::tuple of strings defining column names, types is a list of equal length defining respective
+  // column types types may contain std::optional<?>, which will result in a nullable column, otherwise columns are not
+  // nullable
+  template <typename Names>
+  TableBuilder(size_t chunk_size, const boost::hana::tuple<DataTypes...>& types, const Names& names,
+               opossum::UseMvcc use_mvcc, size_t estimated_rows = 0)
+      : _use_mvcc(use_mvcc), _estimated_rows_per_chunk(std::min(estimated_rows, chunk_size)) {
+    BOOST_HANA_CONSTANT_ASSERT(boost::hana::size(names) == boost::hana::size(types));
+
+    // Iterate over the column types/names and create the columns.
+    auto column_definitions = TableColumnDefinitions{};
+    boost::hana::for_each(boost::hana::zip(names, types), [&](const auto& name_and_type) {
+      auto name = name_and_type[boost::hana::llong_c<0>];
+      auto type = name_and_type[boost::hana::llong_c<1>];
+
+      auto data_type = data_type_from_type<table_builder::get_value_type<decltype(type)>>();
+      auto is_nullable = table_builder::is_optional_v<decltype(type)>;
+
+      column_definitions.emplace_back(name, data_type, is_nullable);
+    });
+    _table = std::make_shared<Table>(column_definitions, TableType::Data, chunk_size, use_mvcc);
+
+    // Reserve some space in the vectors
+    boost::hana::for_each(_value_vectors, [&](auto& values) { values.reserve(_estimated_rows_per_chunk); });
+    boost::hana::for_each(_null_value_vectors, [&](auto& null_values) {
+      if constexpr (std::decay_t<decltype(null_values)>::has_value) {
+        null_values.value().reserve(_estimated_rows_per_chunk);
+      }
+    });
+  }
+
+  std::shared_ptr<Table> finish_table() {
+    if (_current_chunk_row_count() > 0) {
+      _emit_chunk();
+    }
+
+    return _table;
+  }
+
+  // Types == DataTypes, we deduce again to get universal references, so we can call with lvalues
```

Comment:
"so we can call with **rvalues**", no? Don't remember exactly what the problem was, though.

Reply:
No, lvalues is correct. We could not call append_row like this:
`auto customer = call_dbgen_mk<customer_t>(row_idx + 1, mk_cust, TpchTable::Customer);
    customer_builder.append_row(customer.custkey, customer.name, customer.address, customer.nation_code, customer.phone,
                                convert_money(customer.acctbal), customer.mktsegment, customer.comment);`
for tpcds, all in32_t values ahd to be wrapped like `x.append_row(int32_t{customer.age}, ...);` (for tpch it was working for whatever reason).